### DISCUSSION TITLE
how to do close and reconnect

### DISCUSSION THREAD

**Discussion by l-7-l** [gftea/amqprs#94] [Q&A] [answered]
Thanks to the maintainers for the great work.

Can anyone take the time to provide a real world example of dispose and reconnect? I tried to refer to dotnet's [eshoponcontainers](https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/BuildingBlocks/EventBus/EventBusRabbitMQ/DefaultRabbitMQPersistentConnection.cs#L41), but the implementation seems to be different, and there is no complete example on github, any help would be appreciated.

Anyway, thanks again.

**Comment by gftea** [comment, marked as answer]
yes, there is API for listening for error, and user can decide to reconnect 
see discussion on https://github.com/gftea/amqprs/issues/57, if that meets your need

**Reply by l-7-l** [reply]
Thanks

**Reply by l-7-l** [reply]
sorry. i am new to backend. I don't understand， it seems just handle ConnectionShutdown,  how to handle CallbackException and ConnectionBlocked.
why need spawn loop and sleep,   what should I do in a real production environment?
```rust
   let handle1 = tokio::spawn(async move {
        // wait on io failure
        // if return `true`, it means got notification due to network failure, we try to recreate connection
        // if return `false`, the connection may already be closed, or closed later by client or server.
        if connection.listen_network_io_failure().await {
            loop {
                time::sleep(Duration::from_secs(10)).await;

                if let Ok(recover_conn) = Connection::open(&common::build_conn_args()).await {
                    // ... do some recovery...
                    let recover_ch = recover_conn.open_channel(None).await.unwrap();
                    assert!(true == recover_ch.is_open());
                    assert!(true == recover_conn.is_open());
                    info!("Recovery: channel {}", recover_ch);
                    break;
                };
            }
        }
        // here, old connection should be closed no matter due to network failure or closed by server
        assert!(connection.is_open() == false);
    });
```

**Reply by gftea** [reply]
connectionblocked is protocol behavior, you do not need to reconnect
the reconnect is only needed if underlying network conn has unrecoverable error
